Add a guide for migrating data from Postgres to EdgeDB. (#6501)
edgedb/edgedb: docs/guides/datamigrations/postgres.rst

``statuses`` are such tables, so let's look at them:

.. lint-off

.. code-block::

    social=> \d badges
                    Table "public.badges"
       Column    | Type | Collation | Nullable | Default
    -------------+------+-----------+----------+---------
     name        | text |           | not null |
     description | text |           | not null |
    Indexes:
        "badges_pkey" PRIMARY KEY, btree (name)
    Referenced by:
        TABLE "users" CONSTRAINT "users_badge_name_fkey" FOREIGN KEY (badge_name) REFERENCES badges(name)

    social=> \d statuses
                                Table "public.statuses"
     Column |  Type   | Collation | Nullable |               Default
    --------+---------+-----------+----------+--------------------------------------
     id     | integer |           | not null | nextval('statuses_id_seq'::regclass)
     title  | text    |           | not null |
    Indexes:
        "statuses_pkey" PRIMARY KEY, btree (id)
        "statuses_title_key" UNIQUE CONSTRAINT, btree (title)
    Referenced by:
        TABLE "users" CONSTRAINT "users_status_id_fkey" FOREIGN KEY (status_id) REFERENCES statuses(id)

.. lint-on

> The ``badges`` table actually uses the ``name`` of the badge as a primary key
> as opposed to a separate ``id``. So in order to reflect that properly in
> EdgeDB we would have to add :eql:constraint:`exclusive` constraint to this
> property. Meanwhile ``not null`` makes the property ``required``, so we end up
> with a type like this:

Reviewer: ```suggestion
The ``badges`` table uses the ``name`` of the badge as a primary key as opposed
to a separate ``id``. In order to reflect that properly in EdgeDB, we would
have to add an :eql:constraint:`exclusive` constraint to this property.
Meanwhile ``not null`` makes the property ``required``, leaving us with a type
like this:
```

.. code-block:: sql

    type Badge {
        required name: str {
            constraint exclusive;
        }
        required description: str;
    }

> The ``statuses`` table actually has a dedicated ``id`` column in addition to
> ``title``. However, ``id`` in EdgeDB is a :eql:type:`uuid`, whereas in our
> original dataset it is an ``integer``. Luckily, the ``title`` is ``UNIQUE``
> and we might be able to use that to correctly map the data during our
> migration and leave the old ``id`` entirely behind (assuming that no code
> actually depends on it). This leaves us with the following type:

Reviewer: ```suggestion
The ``statuses`` table has a dedicated ``id`` column in addition to ``title``.
However, the automatic ``id`` in EdgeDB is a :eql:type:`uuid`, whereas in our
original dataset it is an ``integer``. Luckily, the ``title`` in the source
data is ``UNIQUE``, and we might be able to use that to correctly map the data
during our migration, leaving the old ``id`` entirely behind (assuming that no
code depends on it). This leaves us with the following type:
```

         Column      |  Type   | Collation | Nullable |              Default
    -----------------+---------+-----------+----------+-----------------------------------
     id              | integer |           | not null | nextval('users_id_seq'::regclass)
     name            | text    |           | not null |
     email           | text    |           | not null |
     password        | text    |           | not null |
     client_settings | jsonb   |           |          |
     badge_name      | text    |           |          |
     status_id       | integer |           |          |
    Indexes:
        "users_pkey" PRIMARY KEY, btree (id)
        "users_email_key" UNIQUE CONSTRAINT, btree (email)
        "users_name_key" UNIQUE CONSTRAINT, btree (name)
    Foreign-key constraints:
        "users_badge_name_fkey" FOREIGN KEY (badge_name) REFERENCES badges(name)
        "users_status_id_fkey" FOREIGN KEY (status_id) REFERENCES statuses(id)
    Referenced by:
        TABLE "bookmarks" CONSTRAINT "bookmarks_user_id_fkey" FOREIGN KEY (user_id) REFERENCES users(id)
        TABLE "posts" CONSTRAINT "posts_user_id_fkey" FOREIGN KEY (user_id) REFERENCES users(id)

.. lint-on

> The ``users`` table also has an ``id`` column, which is not a
> :eql:type:`uuid`. Instead of omitting this information we'll record it as
> ``app_id`` to facilitate the transition. Potentially we still want to
> eventually drop it in favor of the built-in ``id`` from EdgeDB, but we need it
> for now. The column ``client_settings`` would become a :eql:type:`json`
> property. The columns ``badge_name`` and ``status_id`` reference ``badges``
> and ``statuses``, respectively and will become *links* instead of
> *properties*:

Reviewer: ```suggestion
The ``users`` table, like ``statuses``, has an ``id`` column, which is not a
:eql:type:`uuid`. Instead of omitting the ``id`` data, we'll record it as
``app_id`` in EdgeDB to facilitate the transition. We may still want to
eventually drop it in favor of the built-in ``id`` from EdgeDB, but we need it
for now. The column ``client_settings`` would become a :eql:type:`json`
property. The columns ``badge_name`` and ``status_id`` reference ``badges`` and
``statuses`` respectively and will become *links* in EdgeDB instead of
*properties*, even though a property would more closely mirror how they are
stored in Postgres:
```

                ),
            }
        ''',
        line=line.replace(r'\\', '\\'))

Finally, we can deal with the ``bookmarks`` since we've imported both the
users and the posts. The ``bookmarks`` table can be dumped as JSON directly
and then we can write appropriate ``update`` query to add this data to
EdgeDB:

.. code-block:: python

    for line in open('bookmarks.json'):
        client.query('''
            with data := to_json(<str>$line)
            update User
            filter .app_id = <int32>data['user_id']
            set {
                bookmark += (
                    select Post {
                        @note := <str>data['note']
                    }
                    filter .app_id = <int32>data['post_id']
                ),
            }
        ''',
        line=line)

We use the ``+=`` in our ``update`` query to add data incrementally. This way
we don't need to further organize the bookmarks when importing. This approach
also mimics how the bookmarks might be created in the app going forward.

After all the import scripts we end up with data that looks something like
this:

.. lint-off

.. code-block:: edgeql-repl

>     social:edgedb> select User {

Reviewer 2: Yeah, good question. Off the top of my head maybe change to this convention depending on if it's local or cloud? Would show how the instance:database prompt looks now without being overly long.

`local:db>` 

`cloud:db>`

     id     | integer |           | not null | nextval('statuses_id_seq'::regclass)
     title  | text    |           | not null |
    Indexes:
        "statuses_pkey" PRIMARY KEY, btree (id)
        "statuses_title_key" UNIQUE CONSTRAINT, btree (title)
    Referenced by:
        TABLE "users" CONSTRAINT "users_status_id_fkey" FOREIGN KEY (status_id) REFERENCES statuses(id)

.. lint-on

The ``badges`` table actually uses the ``name`` of the badge as a primary key
as opposed to a separate ``id``. So in order to reflect that properly in
EdgeDB we would have to add :eql:constraint:`exclusive` constraint to this
property. Meanwhile ``not null`` makes the property ``required``, so we end up
with a type like this:

.. code-block:: sql

    type Badge {
        required name: str {
            constraint exclusive;
        }
        required description: str;
    }

The ``statuses`` table has a dedicated ``id`` column in addition to ``title``.
However, the automatic ``id`` in EdgeDB is a :eql:type:`uuid`, whereas in our
original dataset it is an ``integer``. Luckily, the ``title`` in the source
data is ``UNIQUE``, and we might be able to use that to correctly map the data

> during our migration, leaving the old ``id`` entirely behind (assuming that no

Reviewer 2: I'm curious about this part vs. the next types - is there a reason for not mapping the old id into a property like app_id (as is done later) or is it just serving as a first example of assuming that you have the id problem solved through some other means?

If the latter then I might rephrase a bit to something like "Let's assume with ``statuses`` here that we already have the old ids correctly mapped during the data, allowing us to leave the old ``id`` entirely behind..." or however you would like to put it.

importing all the data.

We will dump the ``badges`` and ``statuses`` first:

.. lint-off

.. code-block::

    social=> SELECT ROW_TO_JSON(t) FROM badges AS t;
                                    row_to_json
    ---------------------------------------------------------------------------
     {"name":"admin","description":"Superuser who can do anything"}
     {"name":"moderator","description":"User who can edit other user's posts"}
    (2 rows)

    social=> SELECT ROW_TO_JSON(t) FROM statuses AS t;
            row_to_json
    ----------------------------
     {"id":1,"title":"happy"}
     {"id":2,"title":"sad"}
     {"id":3,"title":"excited"}
     {"id":4,"title":"mad"}
    (4 rows)

.. lint-on

These tables can be dumped directly to a file using a ``COPY ... TO
<filename>`` command. We can then read the files and use a simple loop to
import the data into EdgeDB:

.. note::

    When Postgres dumps the JSON data as text files there will be a known
    gotcha causing all the backslashes used to escape characters inside JSON
    string values to be doubled. This is because the *text* format causes
    backslashes themselves to be escaped. This needs to be accounted for when

>     reading the resulting files.

Reviewer 2: Here I would make clear that this is Python code as I could see people misreading this as EdgeQL or some other language. Maybe add this:

`Accomplishing this through the Python client library would look like the following:`
